client/packages/system/src/Immunisation/api/hooks/useImmunisationProgram.ts

```diff
@@ -0,0 +1,174 @@
+  const {
+    mutateAsync: createMutation,
+    isLoading: isCreating,
+    error: createError,
+  } = useCreate();
+
+  const {
+    mutateAsync: updateMutation,
+    isLoading: isUpdating,
+    error: updateError,
+  } = useUpdate(id ?? '', setErrorMessage);
+
+  const draft: DraftImmunisationProgram = data
+    ? { ...defaultDraftImmunisationProgram, ...data, ...patch }
+    : { ...defaultDraftImmunisationProgram, ...patch };
+
+  const updatePatch = (newData: Partial<DraftImmunisationProgram>) => {
+    const newPatch = { ...patch, ...newData };
+    setPatch(newPatch);
+    // Reset error message if user is trying to fix the error
+    setErrorMessage('');
+
+    // Ensures that UI doesn't show in "dirty" state if nothing actually
+    // different from the saved data
+    const updatedData = { ...data, ...newPatch };
+    if (isEqual(data, updatedData)) setIsDirty(false);
+    else setIsDirty(true);
+    return;
+  };
+
+  const create = async () => {
+    const result = await createMutation(draft);
+    setIsDirty(false);
+    return result;
+  };
+
+  const update = async () => {
+    updateMutation(patch);
+    setIsDirty(false);
+  };
+
+  return {
+    query: { data: data, isLoading, error },
+    create: { create, isCreating, createError },
+    update: { update, isUpdating, updateError },
+    draft,
+    errorMessage,
+    isDirty,
+    updatePatch,
+  };
+}
+
+const useGet = (id: string) => {
+  const { api, storeId } = useImmunisationGraphQL();
+
+  const queryFn = async () => {
+    const result = await api.programs({
+      storeId,
+      first: 1,
+      offset: 0,
+      key: ProgramSortFieldInput.Name,
+      desc: false,
+      filter: { id: { equalTo: id } },
+    });
+
+    if (result.programs.__typename === 'ProgramConnector') {
+      return result.programs.nodes[0];
+    }
+  };
+
+  const query = useQuery({
+    queryKey: [PROGRAM, id],
+    queryFn,
+    enabled: id !== '',
+  });
+
+  return query;
+};
+
+const useCreate = () => {
+  const { api, storeId, queryClient } = useImmunisationGraphQL();
+
+  const mutationFn = async ({ name }: DraftImmunisationProgram) => {
+    return await api.insertImmunisationProgram({
+      storeId,
+      input: {
+        id: FnUtils.generateUUID(),
+        name,
+      },
+    });
+  };
+
+  return useMutation({
+    mutationFn,
+    onSuccess: () =>
+      // All Programs need to be re-fetched to include the new one
+      queryClient.invalidateQueries([PROGRAM]),
+  });
+};
+
+const useUpdate = (
+  id: string,
+  setErrorMessage: Dispatch<SetStateAction<string>>
+) => {
+  const { api, storeId, queryClient } = useImmunisationGraphQL();
+  const t = useTranslation('system');
+
+  const mutationFn = async ({ name }: Partial<DraftImmunisationProgram>) => {
+    if (!id) {
+      throw new Error('No ID provided to update Immunisation Program');
+    }
+    if (!name) {
+      throw new Error('No name provided to update Immunisation Program');
```

Comment:
I think we should more gracefully handle someone clicking save with empty input field than throwing to the error boundary 😁 Probably handle in the component, before calling update()? Also should translate these error messages!

Reply:
This was just to make typescript happy.
Save button shouldn't be enabled without these, but yeah could happen in future I guess...

Reply:
Actually you're right, I can trigger it by removing all the text in the name...
Will update.